src/SignalR/clients/csharp/Client/test/UnitTests/HubConnectionTests.cs

```diff
@@ -220,6 +221,104 @@ public async Task SendAsyncCanceledWhenPassedCanceledToken()
         }
     }
 
+    internal class PipeReaderWrapper : PipeReader
+    {
+        private readonly PipeReader _originalPipeReader;
+        private TaskCompletionSource _waitToRead;
+        private readonly object _lock = new object();
+
+        public PipeReaderWrapper(PipeReader pipeReader)
+        {
+            _originalPipeReader = pipeReader;
+            _waitToRead = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed) =>
+            _originalPipeReader.AdvanceTo(consumed);
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined) =>
+            _originalPipeReader.AdvanceTo(consumed, examined);
+
+        public override void CancelPendingRead() =>
+            _originalPipeReader.CancelPendingRead();
+
+        public override void Complete(Exception exception = null) =>
+            _originalPipeReader.Complete(exception);
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            await _waitToRead.Task;
+
+            try
+            {
+                return await _originalPipeReader.ReadAsync(cancellationToken);
+            }
+            finally
+            {
+                lock (_lock)
+                {
+                    _waitToRead = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+                }
+            }
+        }
+
+        public override bool TryRead(out ReadResult result) =>
+            _originalPipeReader.TryRead(out result);
+
+        public void UnblockRead()
+        {
+            lock (_lock)
```

Comment:
This `_lock` isn't doing anything. Grabbing a reference is atomic anyway. If there's a tight race between UnblockRead() and ReadAsync(), that implies the test is flaky anyway.